File: reconstruction/CoilCombination/RefPeak_cc.m
function [Image_cc,Smap] = RefPeak_cc(Image_mc,varargin)
% Reference Peak method for image combination
% Reference:
%   Hall, Emma L., et al. "Methodology for improved detection of low 
%   concentration metabolites in MRS: optimised combination of signals from 
%   multi-element coil arrays." Neuroimage 86 (2014): 35-42.
%
% INPUTS:
%   Image_mc - multi-channel image data. dim:[Nx, Ny, Nz, Ncoil, 1,
%       N_freq, Ndyn]
%       
% OUTPUTS:
%   Image_cc - coil combination Image.
%   Smap - sensitivity map
% (c) 2017-2018 The Regents of the University of California
% All Rights Reserved.
%
% Written by :Xucheng Zhu
%   
%

[Nx, Ny, Nz, Ncoil, Nmap, N_freq, Ndyn] = size(Image_mc);
% sum up data
Image_t = sum(Image_mc,7);
Spectrum_t = squeeze(sum(sum(sum(sum(Image_t,1),2),3),4));
if nargin > 1
    peak_pos = varargin{1};
    fprintf('Manual select metabolite # %d to for RefPeak combination. \n',peak_pos);
else
    [~,peak_pos] = max(squeeze(abs(Spectrum_t)));
    fprintf('Auto select metabolite # %d to for RefPeak combination. \n',peak_pos);
end
Image_ref = Image_t(:,:,:,:,1,peak_pos);

Image_sos = sqrt(abs(sum(Image_ref.*conj(Image_ref),4)));
Smap = Image_ref./repmat(Image_sos,1,1,1,Ncoil,1,1,1);

Image_cc = sum(Image_mc.*repmat(conj(Smap),1,1,1,1,1,N_freq,Ndyn),4);

File: reconstruction/CoilCombination/Python/RefPeak_cc.py
import numpy as np

def ref_peak_cc(Image_mc, peak_pos=None):
    """
    Reference Peak method for image combination.

    Reference:
        Hall, Emma L., et al. "Methodology for improved detection of low 
        concentration metabolites in MRS: optimised combination of signals from 
        multi-element coil arrays." Neuroimage 86 (2014): 35-42.

    Parameters:
        Image_mc (ndarray): Multi-channel image data. Shape: [Nx, Ny, Nz, Ncoil, N_freq, Ndyn]
        peak_pos (int, optional): Index of metabolite peak. If None, selects peak automatically.

    Returns:
        Image_cc (ndarray): Coil-combined image.
        Smap (ndarray): Sensitivity map.

    Written by: Parimal Joshi
    """
    Nx, Ny, Nz, Ncoil, N_freq, Ndyn = Image_mc.shape

    # Sum over dynamics dimension (axis=5)
    Image_t = np.sum(Image_mc, axis=5, keepdims=True)  # shape: [Nx, Ny, Nz, Ncoil, N_freq]
    Spectrum_t = np.squeeze(np.sum(Image_t, axis=(0, 1, 2, 3)))  # shape: [N_freq]

    if peak_pos is None:
        peak_pos = np.argmax(np.abs(Spectrum_t))
        print(f'Auto select metabolite # {peak_pos} for RefPeak combination.')
    else:
        print(f'Manual select metabolite # {peak_pos} for RefPeak combination.')

    Image_ref = Image_t[:, :, :, :, peak_pos]

    Image_sos = np.sqrt(np.sum(np.abs(Image_ref)**2, axis=3, keepdims=True))  
    eps = np.finfo(Image_sos.dtype).eps
    Image_sos[Image_sos == 0] = eps
    # print(f'Image_ref shape: {Image_ref.shape}')
    # print(f'Image_sos shape: {Image_sos.shape}')

    Smap = Image_ref / Image_sos 
    # print(f'Smap shape: {Smap.shape}')
    Smap_exp = np.expand_dims(Smap, axis=(5)) 
    Smap_exp = np.tile(Smap_exp, (1, 1, 1, 1, 1, Ndyn))
    
    Image_cc = np.sum(Image_mc * np.conj(Smap_exp), axis=3)

    return Image_cc, Smap
